### PR TITLE
Generate material library doc by script

### PR DESCRIPTION
**Pull request by xin-flex**
Added script "docs/generate_doc.py" to generate "docs/source/material_library.rst", which is executed in "docs/source/conf.py".

### CODE REVIEW

File: docs/generate_doc.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            if len(ref_string) != 0:
+                f.write('References:\n\n')
+                f.write(ref_string)
+                f.write('\n')
+            
+def main():
+    generate_material_library_doc()
+            
+if __name__ == "__main__":
+    main()
```
Comment (tylerflex) [inline docs/generate_doc.py:170]: add newline to end of file.

**Review by tylerflex — CHANGES_REQUESTED** [review]
Wow, this must have been a lot of work. thanks for including this, it's going to be much improved.

Material library compiles well locally. there are several warnings in the compilation though, seems like dozens of the following warnings

```
/Users/twhughes/Documents/Flexcompute/tidy3d-docs/docs/source/material_library.rst:3: WARNING: Duplicate explicit target name: "[data]".
/Users/twhughes/Documents/Flexcompute/tidy3d-docs/docs/source/material_library.rst:3: WARNING: Duplicate explicit target name: "[doi]".
/Users/twhughes/Documents/Flexcompute/tidy3d-docs/docs/source/material_library.rst:79: WARNING: Blank line required after table.
```

Perhaps some are related to the data / dot links? Let's see if we can somehow get around these warnings?

**Comment by tylerflex** [comment]
One other thing you might want to consider if you have the energy for it is to add a short code example for each variant / material, showing the exact syntax needed to import the medium, for example, for mat "mat" and variant "var" it could look like

```
>>> mat = material_library[mat][var]
```

**Comment by weiliangjin2021** [comment]
Very impressive work!

- Near the top of the page, below "To import a material “mat” with variant “var” as a tidy3d medium:", can we also comment that one can import the default variant with `medium = material_library['mat']`?
- Below `print(material_library['Ag'].variants.keys())`, how about mention that one can access the details of the variants via `material_library['Ag'].variants['var']`?
- Regarding the table, when the number of variants is larger than the number of the reference, it doesn't look so straightforward to find out the reference for the variant. Numbered reference lists and place the number next the variant can be helpful. But maybe it's good enough for now.

**Comment by weiliangjin2021** [comment]
One more question: how are the materials ordered in the material_library doc page? They are not in the alphabetic order.

**Comment by xin-flex** [comment]
> Wow, this must have been a lot of work. thanks for including this, it's going to be much improved.
> 
> Material library compiles well locally. there are several warnings in the compilation though, seems like dozens of the following warnings
> 
> ```
> /Users/twhughes/Documents/Flexcompute/tidy3d-docs/docs/source/material_library.rst:3: WARNING: Duplicate explicit target name: "[data]".
> /Users/twhughes/Documents/Flexcompute/tidy3d-docs/docs/source/material_library.rst:3: WARNING: Duplicate explicit target name: "[doi]".
> /Users/twhughes/Documents/Flexcompute/tidy3d-docs/docs/source/material_library.rst:79: WARNING: Blank line required after table.
> ```
> 
> Perhaps some are related to the data / dot links? Let's see if we can somehow get around these warnings?

This is because multiple material models could reference a same paper. Currently each material's section has its own list of references. One option is to consolidate all references to the end of the page, and reference to it in each material section. I can give it a try to see how it looks.

File: docs/generate_doc.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
+# Generate documentation for Material Library (Python)
+
+import numpy as np
+from tidy3d import material_library as lib
+
+def generate_material_library_doc():
+    
+    # Constant
+    C_0 = 299792458 * 1e6 # um/s
```
[the remark this answers is not in the document]
Comment (xin-flex) [reply docs/generate_doc.py:9]: I can do that

File: docs/generate_doc.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                    lossy_string = 'Yes'
+                else:
+                    lossy_string = 'No'
+
+                # Valid range column
+                if medium.frequency_range is None:
+                    valid_string = 'Not specified'
+                else:
+                    freq = np.array(medium.frequency_range) / 1e12 # THz
+                    wl = C_0 / (freq * 1e12) # um
```
[the remark this answers is not in the document]
Comment (xin-flex) [reply docs/generate_doc.py:89]: got it

File: docs/generate_doc.py (hunk 1 above)
Comment (xin-flex) [reply docs/generate_doc.py:170]: got it

**Comment by xin-flex** [comment]
> add a short code example for each material or variant, showing the exact syntax needed to import the medium, for example, for material "mat" and variant "var" it could look like

Good point!. Let me try to see how it looks.

**Comment by weiliangjin2021** [comment]
> One option is to consolidate all references to the end of the page, and reference to it in each material section. I can give it a try to see how it looks.

Personally, I prefer the current style since each material only has a few references to display. Putting all references to the end makes hard to scroll back and forth. But not sure how others think.

**Comment by xin-flex** [comment]
> Very impressive work!
> 
> * Near the top of the page, below "To import a material “mat” with variant “var” as a tidy3d medium:", can we also comment that one can import the default variant with `medium = material_library['mat']`?
> * Below `print(material_library['Ag'].variants.keys())`, how about mention that one can access the details of the variants via `material_library['Ag'].variants['var']`?
> * Regarding the table, when the number of variants is larger than the number of the reference, it doesn't look so straightforward to find out the reference for the variant. Numbered reference lists and place the number next the variant can be helpful. But maybe it's good enough for now.

Good points! I will add the first too.

For the table and reference, combined w a previous suggestions from Tyler, I think I will try to put all reference to the end of page and use numbered reference. Let me see how it looks.

**Comment by xin-flex** [comment]
> One more question: how are the materials ordered in the material_library doc page? They are not in the alphabetic order.

Probably just the order in the material_library.py.

I can sort them in alphabetical order.

**Comment by xin-flex** [comment]
> > One option is to consolidate all references to the end of the page, and reference to it in each material section. I can give it a try to see how it looks.
> 
> Personally, I prefer the current style since each material only has a few references to display. Putting all references to the end makes hard to scroll back and forth. But not sure how others think.

Ok... let me think about it
Also if Tyler have any thoughts feel free to comment.

**Comment by tylerflex** [comment]
I think i like the original style too. Is there a way to disable sphinx
warnings like this?

**Comment by xin-flex** [comment]
Updates are pushed. Please review. I addressed all comments other than disabling the sphinx warnings of duplicate explicit targets.

**Comment by tylerflex** [comment]
Thanks. I think we should fix the sphinx warnings. Maybe this is worth a shot:
https://github.com/sphinx-doc/sphinx/issues/3921
Basically put two underscores after the link instead of 1 to make it an "anonymous reference".

Having a clean sphinx build isn't totally necessary for the docs compiling well, but it's good to keep a standard that there are no warnings or errors because otherwise they can add up quickly and we might miss some that are actually important to address.

**Comment by xin-flex** [comment]
> Thanks. I think we should fix the sphinx warnings. Maybe this is worth a shot: [sphinx-doc/sphinx#3921](https://github.com/sphinx-doc/sphinx/issues/3921) Basically put two underscores after the link instead of 1 to make it an "anonymous reference".
> 
> Having a clean sphinx build isn't totally necessary for the docs compiling well, but it's good to keep a standard that there are no warnings or errors because otherwise they can add up quickly and we might miss some that are actually important to address.

Good point! Thanks for finding the link! Just fixed it.

**Comment by tylerflex** [comment]
Great!

**Comment by weiliangjin2021** [comment]
Just one last place to change: since we have renamed Rakic1998 in several places to Rakic1998BB, the example at the top `silver = material_library['Ag']['Rakic1998']` need to be updated to Rakic1998BB

**Review by weiliangjin2021 — APPROVED** [review]

**Pull request merged**